feat: add support for serving statically compressed files

chrstntdd opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

resolves #104

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added (types for now - pending input for API)
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

#### Thoughts

I don't have a ton of experience in this testing framework, so I'm sure there's room for improvement on that front as well as accounting for more use-cases. Furthermore I also have a feeling that the implementation of the feature could be done more efficiently, but this is what I could come up with and get the tests passing 🤷 

Definitely open to any and all suggestions to improve what I have so far. 

Thanks!

mcollina commented on test/static.test.js:
```diff
@@ -2283,3 +2283,78 @@ t.test('routes use default errorHandler when fastify.errorHandler is not defined
+  })
+
+  t.strictEqual(response.headers['content-encoding'], 'br')
+  t.strictEqual(response.statusCode, 200)
```
Can you please check that if the precompressed body matches?

chrstntdd replied:
Matches what exactly? The content on disk (similar to the other tests with the top level ` fs.readFileSync` calls)?

mcollina replied:
Yes.

mcollina commented on index.d.ts:
```diff
@@ -41,6 +41,11 @@ export interface FastifyStaticOptions {
+   * @description
+   * Opt-in to looking for pre-compressed files
+   */
+  preCompressed?: boolean;
```
This needs a test for the type

chrstntdd replied:
How can this be accomplished? I see `test/types/index.ts` but unclear about how to write out the test. Is it merely using the new part of the API?

```ts
const options: FastifyStaticOptions = {
  acceptRanges: true,
  cacheControl: true,
  // ...
  preCompressed: false
}
```

mcollina replied:
just do the above

mcollina commented on index.js:
```diff
@@ -38,14 +38,26 @@ function fastifyStatic (fastify, opts, next) {
+
+    if (opts.preCompressed) {
+      if (!checkedExtensions) {
+        checkedExtensions = new Set()
```
I don't understand this logic. Could you explain?

chrstntdd replied:
I've got a recursive call to `pumpSendToReply` within the `error` handler and I wanted to avoid re-creating the data structure where we store what we have already checked for on subsequent calls as we attempt to send other potential compressed versions. I thought about using default parameters, but that seemed too eager since this is an opt-in feature. 

Is it possible to have this structure declared at the plugin level or somewhere else?

mcollina replied:
Can you document it a bit? That's ok.

mcollina commented on index.js:
```diff
@@ -38,14 +38,26 @@ function fastifyStatic (fastify, opts, next) {
+      encodingExtension = checkEncodingHeaders(request.headers, checkedExtensions)
+    }
+
+    const pathnameWithExtension = pathname + `${encodingExtension ? `.${encodingExtension}` : ''}`
```
Can you please avoid the ternary here and move the two values within the `preCompressed` if?

mcollina commented on index.js:
```diff
@@ -234,6 +255,22 @@ function checkRootPathForErrors (fastify, rootPath) {
+
+  const accepted = new Set(headers['accept-encoding'].split(', '))
+
+  let ext; const checkedForBr = checked.has('br')
```
can you please split this in separate lines?

mcollina commented on index.js:
```diff
@@ -234,6 +255,22 @@ function checkRootPathForErrors (fastify, rootPath) {
+function checkEncodingHeaders (headers, checked) {
+  if (!('accept-encoding' in headers)) return
+
+  const accepted = new Set(headers['accept-encoding'].split(', '))
```
I would prefer if you did not use a set but iterate over the array.

mcollina reviewed (changes requested):
Good work! Could you add a test with `wildard: false`? thanks

chrstntdd commented:
> Good work! Could you add a test with `wildard: false`? thanks

How thorough should this check be? Trying it out locally I can duplicate the tests I wrote already and they pass with that option, but not sure if that is the right way to author the tests. How can I write a test that verifies that option works for the cases I have set up without all the duplication?

mcollina commented:
That'd be enough!

chrstntdd commented:
Resolved everything so far. Should we get more eyes on this?

mcollina commented on test/static.test.js:
```diff
@@ -21,6 +21,9 @@ const indexContent = fs.readFileSync('./test/static/index.html').toString('utf8'
 const innerIndex = fs.readFileSync('./test/static/deep/path/for/test/index.html').toString('utf8')
+const allThreeBr = fs.readFileSync('./test/static-pre-compressed/all-three.html.br').toString('utf8')
+const gzipOnly = fs.readFileSync('./test/static-pre-compressed/gzip-only.html.gz').toString('utf8')
+const uncompressedStatic = fs.readFileSync('./test/static-pre-compressed/uncompressed.html').toString('utf8')
```
I won't convert them to utf8, they should be binaries.

mcollina replied:
Ottimitting it would still covert them to utf8. I mean to keep them as buffers.

chrstntdd replied:
Ah. Without that call the tests are failing now. `request.body` is read in the tests as a string. Should I be converting the body into a buffer before checking for the match or does this point to a bug?

mcollina replied:
I think you can use `res.rawPayload` instead. It should be a buffer. https://github.com/fastify/light-my-request/blob/bfffc360ef291df9bfbda9675bf80fc04d534dd7/lib/response.js#L135

mcollina reviewed (approved):
lgtm

chrstntdd commented:
Awesome! Thank you for the patience with the review @mcollina. What are the next steps?

delvedor reviewed (dismissed):
I think it would be useful to support a separate path option for the compressed files. In this way, it might be easier to write automation and remove old files when generating new ones. What do you think?

chrstntdd commented:
@delvedor I think it could be useful, but I am not certain it would be worth the complexity.
How best would that option for a separate directory be supported while also allowing for this simpler case where the compressed file is a sibling?

```ts
const options: FastifyStaticOptions = {
  // ...
  preCompressed?: true | PathLike
}
```

I tried my best to get the same behavior from [`koa-send`](https://github.com/koajs/send/blob/master/index.js#L88).

chrstntdd commented:
I started trying out this branch in a side project and I'm finding that with the new behavior I also have to implement `setHeaders` to set the correct `content-type`. The docs specify that [the default is `application/octet-stream`](https://github.com/fastify/fastify/blob/master/docs/Reply.md#streams), but without `preCompressed` enabled, the `content-type` is correctly set internally. 
Any thoughts about why this might be?

Eomm reviewed (changes requested):
Adding some consideration about the header evaluation.

I would add the separate compressed path feature in a follow-up pr

chrstntdd commented on index.js:
```diff
@@ -234,6 +261,28 @@ function checkRootPathForErrors (fastify, rootPath) {
+  if (!('accept-encoding' in headers)) return
+
+  let ext
+  const accepted = headers['accept-encoding'].split(', ')
```
Would this mean also pulling in `encoding-negotiator` as a dependency?

marcbachmann replied:
The referenced module is quite small, but I don't think that's necessary as most browsers just send the list without priority. For chrome and safari it's `accept-encoding: gzip, deflate, br`

Eomm replied:
> Would this mean also pulling in encoding-negotiator as a dependency?

Yes

I agree that it could seems to much, but I think there is only 2 sentence here:
- support the standard
- not support the standard

To avoid issue, as pointed out, we should in every case add some code like the regxep match, but it seems easier adding the `encoding-negotiator` that is battle-tested for this purpose

(sorry for the late reply)

mcollina commented:
@delvedor @Eomm could you take a look?

Eomm commented:
Could you resolve the conflics?

I think it is necessary to address [this](https://github.com/fastify/fastify-static/pull/158#discussion_r548980565) use case for a safe feature

chrstntdd commented:
Hey folks. I updated this PR with the latest version and added `encoding-negotiator` per the suggestion. Ran into some auto-formatting troubles along the way - apologies for the extra noise but everything seems in order. 

I didn't add any tests with the weighted encoding format since I think all that handled by `encoding-negotiator` itself. Let me know if there's more I can do to get this through.

Thank you all for your patience!

mcollina reviewed (approved):
lgtm

mcollina commented:
@delvedor @Eomm wdyt?

robcresswell commented on README.md:
```diff
@@ -289,6 +289,24 @@ GET .../public/index
 
+#### `preCompressed`
+
+Default: `false`
```
I'm curious why this defaults to false. Is the expectation that people are typically serving precompressed assets with something like nginx in front of Fastify?

As for the general behaviour, what I've observed is that people typically dump all of their assets into a build dir of some sort (as you've shown with `public/` below) and would expect those to be served by whatever is handling their static assets. I guess I'm concerned that this default is unintuitive, and will generate a bunch of "why doesn't this work" issues for the fastify team.

> You may choose to skip compression for smaller files that don't benefit from it.

Usually build processes just avoid compressing these in the first place. I don't think the application should be making this decision, personally. All the webpack compression plugins for example, take a minimum size to compress & save.

This is all very anecdotal, so please feel free to correct me 🤓

robcresswell replied:
Also, just wanted to add, thanks so much for working on this!

mcollina commented:
Could you resolve the conflicts? I'll land as it's landable.

chrstntdd commented:
@mcollina Conflicts resolved! Ready to finally land this before it gets stale again. 😅

mcollina reviewed (approved):
lgtm

mcollina commented:
@delvedor @Eomm any other concerns before landing?
